services/cannon/src/Cannon/API.hs

```diff
@@ -120,24 +131,52 @@ docs (_ ::: url) = do
     return $ responseLBS status200 [jsonContent] doc
 
 push :: UserId ::: ConnId ::: Request -> Cannon Response
-push (user ::: conn ::: req) = do
-    let k = mkKey user conn
+push (user ::: conn ::: req) =
+    singlePush (readBody req) (PushTarget user conn) >>= \case
+        PushStatusOk   -> return empty
+        PushStatusGone -> return $ errorRs status410 "general" "client gone"
+
+-- | Parse the entire list of notifcations and targets, then call 'singlePush' on the each of them
+-- in order.
+bulkpush :: Request -> Cannon Response
+bulkpush req = json <$> (parseJsonBody req >>= bulkpush')
+
+-- | The typed part of 'bulkpush'.
+--
+-- TODO: this sends the push requests to the cannons in sequence ('mapM'), not in parallel
+-- ('mapConcurrently').  it would be nice to fix that, but that requires 'Cannon' to instantiate
+-- 'MonadBaseControl' (or just 'MonadBase'?).
```

Comment:
👌 

> that requires 'Cannon' to instantiate 'MonadBaseControl'

Sounds good 👍

Reply:
2d87aaf42c54fe4d3543e43cf308ea51842c0c73

that just raises further questions...  (-:  but hopefully easier ones.  also, i feel much better about the peformance now, even though i haven't benchmarked it.